lib/topology.ml

```diff
@@ -0,0 +1,275 @@
+(*
+ * Copyright (C) 2019 Citrix Systems Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+module D = Debug.Make (struct
+  let name = "topology"
+end)
+
+open D
+
+module CPUSet = struct
+  include Set.Make (struct
+    type t = int
+
+    let compare (x : int) (y : int) = compare x y
+  end)
+
+  let pp_dump = Fmt.using to_seq Fmt.(Dump.seq int)
+
+  let to_mask t = Array.init (1 + max_elt t) (fun i -> mem i t)
+
+  let all n =
+    n
+    |> ArrayLabels.init ~f:(fun x -> x)
+    |> ArrayLabels.fold_right ~f:add ~init:empty
+end
+
+module NUMAResource = struct
+  type t = {affinity: CPUSet.t; memfree: int64}
+
+  let v ~affinity ~memfree =
+    if Int64.compare memfree 0L < 0 then
+      invalid_arg
+        (Printf.sprintf "NUMAResource: memory cannot be negative: %Ld" memfree) ;
+    {affinity; memfree}
+
+  let empty = {affinity= CPUSet.empty; memfree= 0L}
+
+  let union a b =
+    v
+      ~affinity:(CPUSet.union a.affinity b.affinity)
+      ~memfree:(Int64.add a.memfree b.memfree)
+
+  let min_memory r1 r2 = {r1 with memfree= min r1.memfree r2.memfree}
+
+  let pp_dump =
+    Fmt.(
+      Dump.record
+        [ Dump.field "affinity" (fun t -> t.affinity) CPUSet.pp_dump
+        ; Dump.field "memfree" (fun t -> t.memfree) int64 ])
+end
+
+module NUMARequest = struct
+  type t = {memory: int64; vcpus: int}
+
+  let v ~memory ~vcpus =
+    if Int64.compare memory 0L < 0 then
+      invalid_arg
+        (Printf.sprintf "NUMARequest: memory must be > 0: %Ld" memory) ;
+    if vcpus < 0 then
+      invalid_arg (Printf.sprintf "vcpus cannot be negative: %d" vcpus) ;
+    {memory; vcpus}
+
+  let fits requested available =
+    Int64.compare requested.memory available.NUMAResource.memfree <= 0
+    && CPUSet.(cardinal available.NUMAResource.affinity >= requested.vcpus)
+
+  let shrink a b =
+    v
+      ~memory:(max 0L (Int64.sub a.memory b.NUMAResource.memfree))
+      ~vcpus:(max 0 (a.vcpus - CPUSet.cardinal b.NUMAResource.affinity))
+
+  let pp_dump =
+    Fmt.(
+      Dump.record
+        [ Dump.field "memory" (fun t -> t.memory) int64
+        ; Dump.field "vcpus" (fun t -> t.vcpus) int ])
+end
+
+(** [seq_range a b] is the sequence of numbers between [a, b). *)
+let seq_range a b =
```

Comment:
is there a neat way to extend the `Seq` module in ocaml?

Reply:
There is the OSeq module from the `seq` package which adds some helpers.